### PR TITLE
 Add batching and caching of metric insertions 

### LINKED ISSUE #61: Add batching and caching of metric insertions
Description:
- Add LRU in-memory cache for storing metrics to save in the database
- When there 1000 metrics to insert OR every minute, insert the metrics to the `metrics` table

Estimation: 3

### PR DESCRIPTION
**Pull request by lachhebo**

### CODE REVIEW

**Comment by lachhebo** [comment]
the number of seconds between batch save in db (in the routine) is not parametrized

**Review by frankhillard — CHANGES_REQUESTED** [review]
Which config option should be used ? CacheMaxItems or BlockchainRequestsCacheMaxItems
//before
func NewLRUBlockchainRepository() repository.BlockchainRepository {
	cache, err := lru.New(config.ProxyConfig.Proxy.BlockchainRequestsCacheMaxItems)
//after
func NewCacheBlockchainRepository() repository.BlockchainRepository {
	cache, err := lru.New(config.ProxyConfig.Proxy.CacheMaxItems)

number of seconds of the routine should be in .toml

**Review by frankhillard — APPROVED** [review]

File: internal/proxy/infrastructure/cache/cache_metrics_repository.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		log.Fatal("could not init the LRU cache")
+	}
+
+	return &cacheMetricsRepository{
+		cache: cache,
+	}
+}
+
+func (l cacheMetricsRepository) Add(metric *pkginfradbinputs.MetricsInput) error {
+	newid := l.cache.Len() + 1
```
Comment (louptheron) [inline internal/proxy/infrastructure/cache/cache_metrics_repository.go]: I propose `newid` -> `newKey`

File: internal/proxy/usecases/proxy_usecase.go [hunk 2]
```diff
@@ -21,6 +21,7 @@ type ProxyUsecase struct {
 	metricsRepo      pkgrepository.MetricsRepository
 	projectRepo      pkgrepository.ProjectRepository
 	projectCacheRepo pkgrepository.ProjectRepository
+	lruMetrics       repository.MetricInputRepository
```
Comment (louptheron) [inline internal/proxy/usecases/proxy_usecase.go]: `lruMetrics` -> `metricsCacheRepo`

File: internal/proxy/usecases/proxy_usecase.go [hunk 3]
```diff
@@ -40,13 +41,15 @@ func NewProxyUsecase(
 	proxyRepo repository.BlockchainRepository,
 	metricsRepo pkgrepository.MetricsRepository,
 	projectRepo pkgrepository.ProjectRepository,
-	projectCacheRepo pkgrepository.ProjectRepository) *ProxyUsecase {
+	projectCacheRepo pkgrepository.ProjectRepository,
+	lruMetrics repository.MetricInputRepository) *ProxyUsecase {
```
Comment (louptheron) [inline internal/proxy/usecases/proxy_usecase.go]: lruMetrics -> metricsCacheRepo

File: internal/proxy/usecases/proxy_usecase.go [hunk 4]
```diff
@@ -40,13 +41,15 @@ func NewProxyUsecase(
...
 	return &ProxyUsecase{
 		cacheRepo:        cacheRepo,
 		proxyRepo:        proxyRepo,
 		metricsRepo:      metricsRepo,
 		projectRepo:      projectRepo,
 		projectCacheRepo: projectCacheRepo,
+		lruMetrics:       lruMetrics,
```
Comment (louptheron) [inline internal/proxy/usecases/proxy_usecase.go]: `lruMetrics` -> `metricsCacheRepo`

File: internal/proxy/usecases/proxy_usecase.go [hunk 5]
```diff
@@ -70,6 +73,22 @@ func (p *ProxyUsecase) findInDatabaseIfNotFoundInCache(UUID string) error {
 	return nil
 }
 
+// WriteCachedRequestsRoutine  write metric cache in database
+func (p *ProxyUsecase) WriteCachedRequestsRoutine() {
+	logrus.Info("func WriteCachedRequestsRoutine")
```
Comment (louptheron) [inline internal/proxy/usecases/proxy_usecase.go]: I suggest a log like : "Starting to write cached requests to database"

File: internal/proxy/usecases/proxy_usecase.go [hunk 6]
```diff
@@ -70,6 +73,22 @@ func (p *ProxyUsecase) findInDatabaseIfNotFoundInCache(UUID string) error {
...
+	allRequests, err := p.lruMetrics.GetAll()
+	if err != nil {
+		logrus.Error("could not init the LRU cache")
```
Comment (louptheron) [inline internal/proxy/usecases/proxy_usecase.go]: We need the error, I suggest : `logrus.Errorf("could not get cached metrics: %s", err)`

File: internal/proxy/usecases/proxy_usecase.go [hunk 7]
```diff
@@ -70,6 +73,22 @@ func (p *ProxyUsecase) findInDatabaseIfNotFoundInCache(UUID string) error {
...
+	}
+	logrus.Info("len data", len(allRequests))
```
Comment (louptheron) [inline internal/proxy/usecases/proxy_usecase.go]: I suggest a log like : `logrus.Infof("got %s cached metrics", len(allRequests))`

File: internal/proxy/usecases/proxy_usecase.go [hunk 8]
```diff
@@ -70,6 +73,22 @@ func (p *ProxyUsecase) findInDatabaseIfNotFoundInCache(UUID string) error {
...
+	err = p.metricsRepo.SaveMany(allRequests)
+	if err != nil {
+		logrus.Errorf("could not save metrics in database: %s", err)
+	}
+
```
Comment (louptheron) [inline internal/proxy/usecases/proxy_usecase.go]: I suggest to add a log like : `logrus.Infof("Successfully saved %s cached metrics in database", len(allRequests))`

File: internal/proxy/usecases/proxy_usecase.go [hunk 9]
```diff
@@ -70,6 +73,22 @@ func (p *ProxyUsecase) findInDatabaseIfNotFoundInCache(UUID string) error {
 	return nil
 }
 
+// WriteCachedRequestsRoutine  write metric cache in database
+func (p *ProxyUsecase) WriteCachedRequestsRoutine() {
+	logrus.Info("func WriteCachedRequestsRoutine")
+	allRequests, err := p.lruMetrics.GetAll()
```
Comment (louptheron) [inline internal/proxy/usecases/proxy_usecase.go]: `allRequests` -> `cachedMetrics` ?

File: internal/proxy/usecases/proxy_usecase.go [hunk 10]
```diff
@@ -112,10 +131,28 @@ func (p *ProxyUsecase) Proxy(request *pkgmodel.Request) (string, bool, error) {
 
 func (p *ProxyUsecase) saveMetrics(request *pkgmodel.Request) {
 	metrics := inputs.NewMetricsInput(request, time.Now().UTC())
-	err := p.metricsRepo.Save(&metrics)
+
+	// add to cache
+	err := p.lruMetrics.Add(&metrics)
 	if err != nil {
-		logrus.Errorf("could not save metrics: %s", err)
+		logrus.Error("could not init the LRU cache")
```
Comment (louptheron) [inline internal/proxy/usecases/proxy_usecase.go]: `logrus.Errorf("could not cache the metrics: %s", err)`

File: internal/proxy/usecases/proxy_usecase.go [hunk 11]
```diff
@@ -112,10 +131,28 @@ func (p *ProxyUsecase) Proxy(request *pkgmodel.Request) (string, bool, error) {
...
 	}
+
+	logrus.Info("metric input add to lruMetrics cache")
```
Comment (louptheron) [inline internal/proxy/usecases/proxy_usecase.go]: Same, a phrase would be better

File: pkg/infrastructure/database/postgres_metrics_repository.go [hunk 12]
```diff
@@ -24,16 +26,59 @@ func NewPostgresMetricsRepository(connection *sql.DB) repository.MetricsReposito
...
+	}
+
+	return nil
+}
+
+// insert many metrics
+func (pg postgresMetricsRepository) SaveMany(metricInputs []*inputs.MetricsInput) error {
+	txn, err := pg.connection.Begin()
+	if err != nil {
+		logrus.Error(err)
```
Comment (louptheron) [inline pkg/infrastructure/database/postgres_metrics_repository.go]: As we catch and log the error from the function caller, we don't need this log line anymore

File: pkg/infrastructure/database/postgres_metrics_repository.go [hunk 13]
```diff
@@ -24,16 +26,59 @@ func NewPostgresMetricsRepository(connection *sql.DB) repository.MetricsReposito
...
+		return err
+	}
+
+	stmt, err := txn.Prepare(pq.CopyIn("metrics", "path", "uuid", "remote_address", "date_request"))
+	if err != nil {
+		logrus.Error(err)
```
Comment (louptheron) [inline pkg/infrastructure/database/postgres_metrics_repository.go]: As we catch and log the error from the function caller, we don't need this log line anymore

File: pkg/infrastructure/database/postgres_metrics_repository.go [hunk 14]
```diff
@@ -24,16 +26,59 @@ func NewPostgresMetricsRepository(connection *sql.DB) repository.MetricsReposito
...
+		return err
+	}
+
+	for _, metricInput := range metricInputs {
+		_, err = stmt.Exec(metricInput.Request.Path, metricInput.Request.UUID, metricInput.Request.RemoteAddr, metricInput.Date)
+		if err != nil {
+			logrus.Error(err)
```
Comment (louptheron) [inline pkg/infrastructure/database/postgres_metrics_repository.go]: As we catch and log the error from the function caller, we don't need this log line anymore

File: pkg/infrastructure/database/postgres_metrics_repository.go [hunk 15]
```diff
@@ -24,16 +26,59 @@ func NewPostgresMetricsRepository(connection *sql.DB) repository.MetricsReposito
...
+			return err
+		}
+	}
+
+	_, err = stmt.Exec() // to flush data
+	if err != nil {
+		logrus.Error(err)
```
Comment (louptheron) [inline pkg/infrastructure/database/postgres_metrics_repository.go]: As we catch and log the error from the function caller, we don't need this log line anymore

File: pkg/infrastructure/database/postgres_metrics_repository.go [hunk 16]
```diff
@@ -24,16 +26,59 @@ func NewPostgresMetricsRepository(connection *sql.DB) repository.MetricsReposito
...
+		return err
+	}
+
+	err = stmt.Close()
+	if err != nil {
+		logrus.Error(err)
```
Comment (louptheron) [inline pkg/infrastructure/database/postgres_metrics_repository.go]: As we catch and log the error from the function caller, we don't need this log line anymore

File: pkg/infrastructure/database/postgres_metrics_repository.go [hunk 17]
```diff
@@ -24,16 +26,59 @@ func NewPostgresMetricsRepository(connection *sql.DB) repository.MetricsReposito
...
+		return err
+	}
 
+	err = txn.Commit()
 	if err != nil {
-		return fmt.Errorf("could not insert metrics for UUID %s: %s", metrics.Request.UUID, err)
+		logrus.Error(err)
```
Comment (louptheron) [inline pkg/infrastructure/database/postgres_metrics_repository.go]: As we catch and log the error from the function caller, we don't need this log line anymore

**Review by louptheron — CHANGES_REQUESTED** [review]
Thx!
I suggested some change mainly around logging.
Did you tested with with docker-compose and some `curl` requests ?

File: internal/proxy/usecases/proxy_usecase.go (hunk 2 above)
Comment (frankhillard) [reply internal/proxy/usecases/proxy_usecase.go]: renamed

File: internal/proxy/usecases/proxy_usecase.go (hunk 3 above)
Comment (frankhillard) [reply internal/proxy/usecases/proxy_usecase.go]: renamed

File: pkg/infrastructure/database/postgres_metrics_repository.go (hunk 13 above)
Comment (frankhillard) [reply pkg/infrastructure/database/postgres_metrics_repository.go]: comment removed

File: pkg/infrastructure/database/postgres_metrics_repository.go (hunk 14 above)
Comment (frankhillard) [reply pkg/infrastructure/database/postgres_metrics_repository.go]: comment removed

File: pkg/infrastructure/database/postgres_metrics_repository.go (hunk 15 above)
Comment (frankhillard) [reply pkg/infrastructure/database/postgres_metrics_repository.go]: comment removed

File: pkg/infrastructure/database/postgres_metrics_repository.go (hunk 17 above)
Comment (frankhillard) [reply pkg/infrastructure/database/postgres_metrics_repository.go]: comment removed

File: internal/proxy/usecases/proxy_usecase.go (hunk 5 above)
Comment (frankhillard) [reply internal/proxy/usecases/proxy_usecase.go]: replaced

File: pkg/infrastructure/database/postgres_metrics_repository.go (hunk 16 above)
Comment (frankhillard) [reply pkg/infrastructure/database/postgres_metrics_repository.go]: log removed

File: internal/proxy/infrastructure/cache/cache_metrics_repository.go (hunk 1 above)
Comment (frankhillard) [reply internal/proxy/infrastructure/cache/cache_metrics_repository.go]: ok done

**Comment by frankhillard** [comment]
> Thx!
> I suggested some change mainly around logging.
> Did you tested with with docker-compose and some `curl` requests ?
Answer: Yes
We tested the go routine executed every minute
We tested the cache limit (verifying database write when limit reach)

**Review by louptheron — APPROVED** [review]

**Pull request merged**